src/futures/digital.rs

```diff
@@ -0,0 +1,57 @@
+//! Asynchronous digital I/O
+
+use core::future::Future;
+
+use crate::blocking::digital::InputPin;
+
+/// Asynchronously wait for a pin to become high or low.
+///
+/// # Examples
+/// ```rust
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// /// Asynchronously wait until the `ready_pin` becomes high.
+/// async fn wait_until_ready<P>(ready_pin: &P)
+/// where
+///     P: AsyncInputPin,
+/// {
+///     ready_pin
+///         .until_high()
+///         .await
+///         .expect("failed to await input pin")
+/// }
+/// ```
+///
+/// ```rust,ignore
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// # use embedded_hal::futures::delay::Delay;
+/// use core::time::Duration;
+///
+/// /// Wait until the `ready_pin` is high or timeout after 1 millisecond.
+/// /// Returns true is the pin became high or false if it timed-out.
+/// async fn wait_until_ready_or_timeout<P, D>(ready_pin: &P, delay: &mut D) -> bool
+/// where
+///     P: AsyncInputPin,
+///     D: Delay,
+/// {
+///     futures::select_biased! {
+///         x => ready_pin.until_high() => {
+///             x.expect("failed to await input pin");
+///             true
+///         },
+///         _ => delay.delay(Duration::from_millis(1)) => false, // ignore the error
+///     }
+/// }
+/// ```
+pub trait AsyncInputPin: InputPin {
+    /// The future returned by the `until_high` function.
+    type UntilHighFuture<'a>: Future<Output=Result<(), Self::Error>> + 'a;
+
+    /// The future returned by the `until_low` function.
+    type UntilLowFuture<'a>: Future<Output=Result<(), Self::Error>> + 'a;
+
+    /// Returns a future that resolves when this pin becomes high.
```

Comment:
Perhaps clarify that it waits for the pin to *be* high, not to undergo a low-to-high *transition*. ie if the pin is already high it returns immediately.

Reply:
Good call.

Reply:
One worry I have here is that the future might need to have some state here. On the first poll call, it'll need to check, but it shouldn't after that. It sounds a little error-prone.

Reply:
Some hardware allows you to directly wait for high/low (nrf's GPIOTE). Some doesn't (stm32's EXTI) so the impl would have to do extra check yup, and doing them without race conditions is tricky.

However I think it's a good thing to force HALs to deal with that, otherwise the burden shifts to the end user. External chip interrupts are often "active low" and you want `wait_for_low` to return immediately if it's already low. If it waits for a falling edge it's super easy to get stuck in a case where the line is low and you're stuck waiting forever for a falling edge that never comes.

Reply:
Thanks for the explanation.